### PR TITLE
[Opt] optimize the erase performance 

### PR DESCRIPTION
**Pull request by rhdong**
- [Opt] optimize the erase performance 
- [Benchmark] add the benchmark for erase

### CODE REVIEW

File: include/merlin/core_kernels.cuh [hunk 1]
```diff
@@ -604,25 +606,35 @@ __global__ void upsert_kernel_with_io(
...
+        } else {
+           src_lane = __ffs(reclaim_vote) - 1;
+        }
         key_pos = (start_idx + tile_offset + src_lane) & (bucket_max_size - 1);
         local_size = buckets_size[bkt_idx];
         if (rank == src_lane) {
           bucket->keys[key_pos] = insert_key;
           bucket->metas[key_pos].val = metas[key_idx];
-          if (current_key == EMPTY_KEY) {
+          if (current_key == EMPTY_KEY || current_key == RECLAIM_KEY) {
```
Comment (Lifann) [inline include/merlin/core_kernels.cuh:637]: Can merge same comparisons statement.

File: include/merlin/core_kernels.cuh [hunk 2]
```diff
@@ -604,25 +606,35 @@ __global__ void upsert_kernel_with_io(
 
     int src_lane = -1;
 
+    uint32_t tile_offset = 0;
+    size_t key_offset;
```
Comment (Lifann) [inline include/merlin/core_kernels.cuh:610]: size_t key_offset = 0;

File: include/merlin/core_kernels.cuh [hunk 3]
```diff
@@ -604,25 +606,35 @@ __global__ void upsert_kernel_with_io(
...
+    K current_key;
```
Comment (Lifann) [inline include/merlin/core_kernels.cuh:611]: K current_key = 0;

File: include/merlin/core_kernels.cuh [hunk 4]
```diff
@@ -680,27 +714,37 @@ __global__ void upsert_kernel_with_io(
...
```
Comment (Lifann) [inline include/merlin/core_kernels.cuh:719]: Same as above

File: include/merlin/types.cuh [hunk 5]
```diff
@@ -28,6 +28,7 @@ struct Meta {
 };
 
 constexpr uint64_t EMPTY_KEY = 0xFFFFFFFFFFFFFFFF;
+constexpr uint64_t RECLAIM_KEY = 0xFFFFFFFFFFFFFFFE;
```
Comment (Lifann) [inline include/merlin/types.cuh:31]: What's the purpose of RECLAIM_KEY. Seems it only changes value when replacement happens in full bucket.

File: include/merlin/core_kernels.cuh [hunk 6]
```diff
@@ -841,27 +939,36 @@ __global__ void upsert_kernel(const Table<K, V, M, DIM>* __restrict table,
 
     int src_lane = -1;
 
+    uint32_t tile_offset = 0;
+    size_t key_offset;
```
Comment (Lifann) [inline include/merlin/core_kernels.cuh:943]: size_t key_offset = 0;

File: include/merlin/core_kernels.cuh [hunk 7]
```diff
@@ -841,27 +939,36 @@ __global__ void upsert_kernel(const Table<K, V, M, DIM>* __restrict table,
...
+    K current_key;
```
Comment (Lifann) [inline include/merlin/core_kernels.cuh:944]: K current_key = 0;

Comment (rhdong) [reply include/merlin/core_kernels.cuh:944]: Accept.

File: include/merlin/types.cuh (hunk 5 above)
Comment (rhdong) [reply include/merlin/types.cuh:31]: It's used in the `core.cuh`

**Review by Lifann — DISMISSED** [review]
LGTM

**Pull request merged**